secure/storage/src/storage.rs

```diff
@@ -10,68 +14,310 @@ use libra_crypto::{
+            BoxedStorage::OnDiskStorage(_v) => {
+                Err("Tried to convert variant OnDiskStorage to VaultStorage")
+            }
+        }
+    }
+}
+impl TryInto<InMemoryStorage> for BoxedStorage {
+    type Error = &'static str;
+    fn try_into(self) -> ::core::result::Result<InMemoryStorage, Self::Error> {
+        match self {
+            BoxedStorage::InMemoryStorage(v) => Ok(v),
+            BoxedStorage::GitHubStorage(_v) => {
+                Err("Tried to convert variant GitHubStorage to InMemoryStorage")
+            }
+            BoxedStorage::VaultStorage(_v) => {
+                Err("Tried to convert variant VaultStorage to InMemoryStorage")
+            }
+            BoxedStorage::NamespacedStorage(_v) => {
+                Err("Tried to convert variant NamespacedStorage to InMemoryStorage")
+            }
+            BoxedStorage::OnDiskStorage(_v) => {
+                Err("Tried to convert variant OnDiskStorage to InMemoryStorage")
+            }
+        }
+    }
+}
+impl TryInto<NamespacedStorage> for BoxedStorage {
+    type Error = &'static str;
+    fn try_into(self) -> ::core::result::Result<NamespacedStorage, Self::Error> {
+        match self {
+            BoxedStorage::NamespacedStorage(v) => Ok(v),
+            BoxedStorage::GitHubStorage(_v) => {
+                Err("Tried to convert variant GitHubStorage to NamespacedStorage")
+            }
+            BoxedStorage::VaultStorage(_v) => {
+                Err("Tried to convert variant VaultStorage to NamespacedStorage")
+            }
+            BoxedStorage::InMemoryStorage(_v) => {
+                Err("Tried to convert variant InMemoryStorage to NamespacedStorage")
+            }
+            BoxedStorage::OnDiskStorage(_v) => {
+                Err("Tried to convert variant OnDiskStorage to NamespacedStorage")
+            }
+        }
+    }
+}
+impl TryInto<OnDiskStorage> for BoxedStorage {
+    type Error = &'static str;
+    fn try_into(self) -> ::core::result::Result<OnDiskStorage, Self::Error> {
+        match self {
+            BoxedStorage::OnDiskStorage(v) => Ok(v),
+            BoxedStorage::GitHubStorage(_v) => {
+                Err("Tried to convert variant GitHubStorage to OnDiskStorage")
+            }
+            BoxedStorage::VaultStorage(_v) => {
+                Err("Tried to convert variant VaultStorage to OnDiskStorage")
+            }
+            BoxedStorage::InMemoryStorage(_v) => {
+                Err("Tried to convert variant InMemoryStorage to OnDiskStorage")
+            }
+            BoxedStorage::NamespacedStorage(_v) => {
+                Err("Tried to convert variant NamespacedStorage to OnDiskStorage")
+            }
+        }
+    }
+}
 
-impl KVStorage for BoxStorage {
+impl KVStorage for BoxedStorage {
+    #[inline]
     fn available(&self) -> Result<(), Error> {
-        self.0.available()
+        match self {
+            BoxedStorage::GitHubStorage(inner) => KVStorage::available(inner),
```

Comment:
what's the advantage of this format versus
```inner.available()```

Reply:
Just fully-qualified code that's impervious to any wandering method the underlying concrete type (e.g. GithubStorage ...) might implement.